presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -189,21 +195,27 @@ public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Voi
 
             private Expression rewriteExpressionWithResolvedName(Expression node)
             {
-                Optional<Integer> fieldIndex = analysis.getFieldIndex(node);
-                checkState(fieldIndex.isPresent(), "No field mapping for node '%s'", node);
-
-                Symbol symbol = rewriteBase.getSymbol(fieldIndex.get());
-                checkState(symbol != null, "No symbol mapping for node '%s' (%s)", node, fieldIndex.get());
-                Expression rewrittenExpression = symbol.toSymbolReference();
-
-                return coerceIfNecessary(node, rewrittenExpression);
+                Optional<Symbol> symbol = rewriteBase.getSymbol(node);
+                if (symbol.isPresent()) {
+                    return coerceIfNecessary(node, symbol.get().toSymbolReference());
+                }
+                return node;
             }
 
             @Override
             public Expression rewriteDereferenceExpression(DereferenceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
             {
-                if (analysis.getFieldIndex(node).isPresent()) {
-                    return rewriteExpressionWithResolvedName(node);
+                Optional<ResolvedField> resolvedField = rewriteBase.getScope().tryResolveField(node);
+                if (resolvedField.isPresent()) {
+                    if (resolvedField.get().isLocal()) {
+                        Optional<Symbol> symbol = rewriteBase.getSymbol(node);
+                        checkState(symbol.isPresent(), "No symbol mapping for node '%s'", node);
+                        Expression rewrittenExpression = symbol.get().toSymbolReference();
```

Comment:
Done and I tried to apply the second comment but I went more complicated than it was before.